File: apps/desktop-shell/src/lib/stores/jobs.ts
/// The Activity/Jobs feed (job-progress-surface.md): the shell-owned aggregator of
/// long-running background work - file operations, package installs, model
/// downloads, transfers - shown as a dedicated zone at the top of the notifications
/// popover. A job is live, progressing, and cancelable; a notification is a past
/// event. Closes the acute gap that the file manager reports no progress today.
///
/// Mock-vs-live: fixture-backed. The JobView feed (the notification-daemon extended
/// into a KDE-JobViewV3-mirror job server + the producers reporting progress) is a
/// coder seam; the `list_jobs` query + cancel/pause/resume commands + the event feed
/// are not built. Under vite the store serves a fixture set so the zone renders. The
/// shell owns the threshold/min-dwell visibility (a job shows once it passes ~1.5s).

import { writable } from "svelte/store";
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { tauriAvailable } from "$lib/tauri";

/// A job's lifecycle state (mirrors the JobView state enum).
export type JobState =
  | "running"
  | "paused"
  | "impeded"
  | "error_recoverable"
  | "error_fatal"
  | "done";

/// One real-unit metric ("84 of 240 files"). The consumer derives the sentence;
/// the feed never pre-bakes a percentage.
export interface JobMetric {
  processed: number;
  total: number;
  unit: string;
}

/// One item inside a composite job, for the expandable per-item list (never hide
/// the per-file names behind a lone aggregate bar).
export interface JobItem {
  name: string;
  done: boolean;
}


/// A long-running operation, as the zone renders it.
export interface Job {
  id: string;
  title: string;
  /// The attested producer app id + a friendly label.
  appId: string;
  appLabel: string;
  /// A monotonic 0..1 fraction (never backwards), kept separate from the ETA.
  fraction: number;
  state: JobState;
  metrics: JobMetric[];
  /// A coarse ETA ("about 3 minutes"), never false hh:mm:ss precision.
  etaText?: string;
  killable: boolean;
  suspendable: boolean;
  /// A message for the non-running error/impeded states.
  error?: string;
  /// The host a network job reaches (no-silent-egress transparency).
  egressHost?: string;
  /// The entries this job works through, with the finished ones marked.
  ///
  /// The zone expands its aggregate bar into these: showing only a total that
  /// hides the file names is the mistake the plan names outright. Which are done
  /// is DERIVED in the shell backend from the names and the count, not carried
  /// per item - two copies of one fact eventually disagree.
  items: JobItem[];
  /// When the producer started it, epoch micros.
  ///
  /// The zone owns the visibility threshold - a job shows once it has run long
  /// enough to be worth a row - and this is what that decision is made from. The
  /// daemon deliberately does not decide it: a threshold in the producer would
  /// be every producer inventing its own idea of "worth showing".
  startedAt: number;
}

// i18n-foreign: a job's title is written by whichever daemon is doing the work
// and names the user's own files - "Copying 240 photos to USB" is a sentence
// about their photos, not a label of ours. The fixture stands in for that, so it
// is not ours to translate either; making the daemons send an id instead is the
// same design question the consent summary raises.
const MOCK_JOBS: Job[] = [
  {
    id: "fm-copy",
    startedAt: 1000000,
    title: "Copying 240 photos to USB",
    appId: "org.arlen.files",
    appLabel: "Files",
    fraction: 0.35,
    state: "running",
    metrics: [
      { processed: 84, total: 240, unit: "files" },
      { processed: 120, total: 340, unit: "MB" },
    ],
    etaText: "about 2 minutes",
    killable: true,
    suspendable: true,
    items: [
      { name: "IMG_2103.jpg", done: true },
      { name: "IMG_2104.jpg", done: true },
      { name: "IMG_2105.jpg", done: false },
      { name: "IMG_2106.jpg", done: false },
    ],
  },
  {
    id: "model-pull",
    items: [],
    startedAt: 2000000,
    title: "Downloading the language model",
    appId: "org.arlen.assistant",
    appLabel: "Assistant",
    fraction: 0.22,
    state: "running",
    metrics: [{ processed: 1300, total: 5900, unit: "MB" }],
    etaText: "about 6 minutes",
    killable: true,
    suspendable: true,
    egressHost: "huggingface.co",
  },
  {
    id: "transfer",
    items: [],
    startedAt: 3000000,
    title: "Sending files to your laptop",
    appId: "org.arlen.files",
    appLabel: "Files",
    fraction: 0.5,
    state: "paused",
    metrics: [{ processed: 5, total: 12, unit: "files" }],
    killable: true,
    suspendable: true,
  },
  {
    id: "convert",
    items: [],
    startedAt: 4000000,
    title: "Converting clip.mp4",
    appId: "org.arlen.media",
    appLabel: "Media",
    fraction: 0.8,
    state: "error_recoverable",
    metrics: [{ processed: 48, total: 60, unit: "seconds" }],
    error: "Ran out of disk space. Free some room and retry.",
    killable: true,
    suspendable: false,
  },
  {
    id: "fm-done",
    items: [],
    startedAt: 5000000,
    title: "Copied 18 files to Documents",
    appId: "org.arlen.files",
    appLabel: "Files",
    fraction: 1,
    state: "done",
    metrics: [{ processed: 18, total: 18, unit: "files" }],
    killable: false,
    suspendable: false,
  },
];

/// The jobs on screen now (fixture until the JobView feed lands).
export const jobs = writable<Job[]>([]);

/// True while the feed is the MOCK, not real background work. The zone shows a
/// titled job with real-unit metrics ("84 of 240 files"), an ETA, per-file names
/// and a Cancel button - unlabelled it reads as a copy actually in flight, and a
/// user could cancel it, or unplug the drive believing it is 35% done.
export const mocked = writable(false);

/// The last action failure, for the zone to show. Empty when all is well.
/// The keys this store may hold. A UNION rather than `string`, because a
/// `string` is also what `String(e)` and an English sentence written here both
/// are - and both were written into stores exactly like this one tonight. Naming
/// them makes a wrong write a compile error rather than something a check has to
/// find afterwards.
export type JobMessage =
  | ""
  | "sh.jobs.unavailable"
  | "sh.job.notCancelled"
  | "sh.job.notPaused"
  | "sh.job.notResumed";

/// The message KEY of the last refusal, or "" for none.
///
/// A key rather than a resolved sentence, so the line follows a locale change
/// while it is on screen: `get(t)(...)` freezes the wording at the moment of
/// failure, and this line stays up until the next action replaces it.
export const lastError = writable<JobMessage>("");

/// Load the current jobs. Live: `list_jobs` + the event feed; fixture under vite.
export async function pollJobs(): Promise<void> {
  try {
    jobs.set(await invoke<Job[]>("list_jobs"));
    mocked.set(false);
  } catch {
    if (!tauriAvailable) {
      jobs.set(MOCK_JOBS);
      mocked.set(true);
      return;
    }
    // A real session: show no jobs rather than invented ones. These rows carry
    // actions - cancel, retry - so a fabricated job is a button that acts on
    // something that does not exist, and the zone's own `lastError` is where the
    // failure belongs.
    jobs.set([]);
    mocked.set(false);
    lastError.set("sh.jobs.unavailable");
  }
}

/// Follow the daemon's live feed.
///
/// `pollJobs` is the snapshot a zone opens with; this is what keeps it moving.
/// The two carry the SAME row - the shell builds it once from the wire and sends
/// it both ways - so a job cannot read one way in the list and another as it
/// updates.
///
/// A job arrives with `removed` when it finished or was cancelled, and it leaves
/// the feed at that point: the transient "done" receipt is the zone's to show,
/// and a store that kept finished rows would make the list a history rather than
/// a picture of what is running.
export async function watchJobs(): Promise<UnlistenFn | null> {
  if (!tauriAvailable) return null;
  return listen<{ job: Job; removed: boolean }>("notification:job", (event) => {
    const { job, removed } = event.payload;
    jobs.update((list) => {
      const rest = list.filter((j) => j.id !== job.id);
      // Oldest first, the same order `list_jobs` returns, so the snapshot and
      // the feed cannot disagree about where a row sits. A row that moves as its
      // progress changes is a row nobody can click.
      return removed
        ? rest
        : [...rest, job].sort((a, b) => a.startedAt - b.startedAt || a.id.localeCompare(b.id));
    });
  });
}

/// Drive one job action optimistically, then reconcile with the daemon.
///
/// A REAL refusal restores the previous feed and says so. Swallowing it would
/// report a cancelled copy that is still running - the same false confirmation of
/// a destructive action the task manager had. Without the runtime there is no
/// daemon to refuse, so the optimistic mock stands.
/// `failure` is a message KEY, not a sentence.
///
/// It used to be an English string written here - "Could not cancel that job" -
/// with `String(e)` appended, and the zone rendered the pair verbatim. So the
/// only line a person sees when a cancel is refused was written in a TypeScript
/// store where no catalogue can reach it, followed by whatever the daemon
/// formatted. The born-translatable lint reads a prose literal assigned to an
/// `*error`/`*message` NAME; passed as an argument called `failure`, this slipped
/// past it.
async function driveJob(
  id: string,
  apply: (list: Job[]) => Job[],
  cmd: string,
  failure: JobMessage,
): Promise<void> {
  let previous: Job[] = [];
  jobs.update((list) => {
    previous = list;
    return apply(list);
  });
  try {
    await invoke(cmd, { id });
  } catch (e) {
    if (tauriAvailable) {
      jobs.set(previous);
      // The daemon's own words name a command and carry an errno. They go where
      // whoever debugs this will read them; the zone gets the sentence.
      console.warn(`shell: ${cmd} refused`, e);
      lastError.set(failure);
    }
  }
}

/// Cancel a job (a clean cancel, per the Killable flag). Live: `cancel_job`.
export async function cancelJob(id: string): Promise<void> {
  await driveJob(id, (l) => l.filter((j) => j.id !== id), "cancel_job", "sh.job.notCancelled");
}

/// Pause a suspendable job. Live: `pause_job`.
export async function pauseJob(id: string): Promise<void> {
  await driveJob(
    id,
    (l) => l.map((j) => (j.id === id ? { ...j, state: "paused" } : j)),
    "pause_job",
    "sh.job.notPaused",
  );
}

/// Resume a paused job. Live: `resume_job`.
export async function resumeJob(id: string): Promise<void> {
  await driveJob(
    id,
    (l) => l.map((j) => (j.id === id ? { ...j, state: "running" } : j)),
    "resume_job",
    "sh.job.notResumed",
  );
}

File: apps/desktop-shell/src-tauri/src/jobs.rs
//! The Activity/Jobs zone's backing state (`job-progress-surface.md`).
//!
//! The notification daemon hosts the JobView server and pushes every register,
//! update and finish down the socket this shell already holds; the client
//! forwards each one here AND to the frontend. This module is the shell's live
//! set, so a store that loads after a job started has something to render rather
//! than waiting for that job's next tick - a paused download might not send one
//! for a long time.
//!
//! THE TRANSLATION LIVES HERE, in one place, and it is not cosmetic. The wire
//! spells the error states `error-recoverable` and `error-fatal`, the KDE
//! JobViewV3 tokens the daemon mirrors; the store declares them
//! `error_recoverable` and `error_fatal`. Forwarding the wire spelling verbatim
//! meant a job that failed arrived in a state the zone has no branch for, so the
//! one moment the surface exists to explain rendered as nothing. The shell is
//! the translator between the daemon's vocabulary and the surface's, and this is
//! the only place either name appears.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use serde::Serialize;

/// One real-unit metric, as the zone renders it ("84 of 240 files").
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct JobMetric {
    pub processed: u64,
    pub total: u64,
    pub unit: String,
}

/// One job in the shape the store declares.
#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct JobRow {
    pub id: String,
    pub title: String,
    pub app_id: String,
    /// What to call the producing app. The id until the shell can do better:
    /// showing `org.arlen.files` is plain, and inventing a prettier name for an
    /// app nobody looked up would be worse.
    pub app_label: String,
    pub fraction: f64,
    pub state: String,
    pub metrics: Vec<JobMetric>,
    pub killable: bool,
    pub suspendable: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub egress_host: Option<String>,
    /// When it began, epoch micros. The SHELL owns the visibility threshold, so
    /// this travels rather than a daemon-side decision about what is worth
    /// showing.
    pub started_at: u64,
}

/// The store's spelling of a wire state token.
///
/// An unknown token is carried through rather than dropped or rounded to
/// `running`: a job in a state this build does not know about is still a job,
/// and claiming it is running would be the surface saying something it was not
/// told.
#[must_use]
pub fn state_token(wire: &str) -> String {
    match wire {
        "error-recoverable" => "error_recoverable".to_string(),
        "error-fatal" => "error_fatal".to_string(),
        other => other.to_string(),
    }
}

/// The zone's row for one job update.
#[must_use]
pub fn row_from_update(
    id: u64,
    app_id: &str,
    title: &str,
    state: &str,
    state_message: &str,
    unit: &str,
    processed: u64,
    determinate: bool,
    total: u64,
    fraction: f64,
    killable: bool,
    suspendable: bool,
    started_at: u64,
    egress_host: &str,
) -> JobRow {
    JobRow {
        id: id.to_string(),
        title: title.to_string(),
        app_id: app_id.to_string(),
        app_label: app_id.to_string(),
        fraction,
        state: state_token(state),
        // A job that does not know its total yet reports no metric at all. The
        // alternative - "84 of 0 files" - is a sentence about a total nobody has
        // counted, and the zone draws an indeterminate bar from the flag instead.
        metrics: if determinate {
            vec![JobMetric {
                processed,
                total,
                unit: unit.to_string(),
            }]
        } else {
            Vec::new()
        },
        killable,
        suspendable,
        error: (!state_message.is_empty()).then(|| state_message.to_string()),
        egress_host: (!egress_host.is_empty()).then(|| egress_host.to_string()),
        started_at,
    }
}

/// The jobs running right now, by id.
pub type LiveJobs = Arc<Mutex<HashMap<u64, JobRow>>>;

/// A fresh, empty live set.
#[must_use]
pub fn new_live_jobs() -> LiveJobs {
    Arc::new(Mutex::new(HashMap::new()))
}

/// Fold one update into the live set. `removed` takes the job out.
pub fn apply(live: &LiveJobs, id: u64, row: JobRow, removed: bool) {
    let Ok(mut map) = live.lock() else {
        return;
    };
    if removed {
        map.remove(&id);
    } else {
        map.insert(id, row);
    }
}

/// The jobs running right now, oldest first.
///
/// Oldest first because the zone is a list somebody reads top down and a job
/// that jumps position as its progress changes is a list nobody can click in.
#[tauri::command]
pub fn list_jobs(live: tauri::State<'_, LiveJobs>) -> Vec<JobRow> {
    let Ok(map) = live.lock() else {
        return Vec::new();
    };
    let mut rows: Vec<JobRow> = map.values().cloned().collect();
    rows.sort_by(|a, b| a.started_at.cmp(&b.started_at).then(a.id.cmp(&b.id)));
    rows
}

#[cfg(test)]
mod tests {
    use super::*;

    fn row(id: u64, state: &str, started_at: u64) -> JobRow {
        row_from_update(
            id,
            "org.arlen.files",
            "Copying 240 photos to USB",
            state,
            if state == "running" { "" } else { "The disk is full" },
            "files",
            84,
            true,
            240,
            0.35,
            true,
            false,
            started_at,
            "",
        )
    }

    #[test]
    fn the_error_states_arrive_in_the_spelling_the_zone_branches_on() {
        assert_eq!(state_token("error-recoverable"), "error_recoverable");
        assert_eq!(state_token("error-fatal"), "error_fatal");
        assert_eq!(state_token("running"), "running");
    }

    #[test]
    fn a_state_this_build_does_not_know_is_carried_not_guessed() {
        assert_eq!(state_token("throttled"), "throttled");
    }

    #[test]
    fn a_running_job_carries_no_error_line() {
        let r = row(1, "running", 10);
        assert_eq!(r.error, None);
        assert_eq!(r.state, "running");
        assert_eq!(r.metrics, vec![JobMetric { processed: 84, total: 240, unit: "files".into() }]);
    }

    #[test]
    fn a_stalled_job_says_why() {
        let r = row(1, "impeded", 10);
        assert_eq!(r.error.as_deref(), Some("The disk is full"));
    }

    #[test]
    fn a_job_with_no_total_yet_reports_no_metric_rather_than_a_zero() {
        let r = row_from_update(1, "a", "t", "running", "", "bytes", 500, false, 0, 0.0, false, false, 1, "");
        assert!(r.metrics.is_empty(), "84 of 0 is not a sentence about anything");
    }

    #[test]
    fn a_networked_job_names_the_host_it_reaches() {
        let r = row_from_update(1, "a", "t", "running", "", "bytes", 1, true, 2, 0.5, false, false, 1, "cdn.example");
        assert_eq!(r.egress_host.as_deref(), Some("cdn.example"));
        let local = row(2, "running", 1);
        assert_eq!(local.egress_host, None, "a local job claims no destination");
    }

    #[test]
    fn an_update_replaces_its_job_and_a_finish_removes_it() {
        let live = new_live_jobs();
        apply(&live, 1, row(1, "running", 10), false);
        apply(&live, 1, row(1, "paused", 10), false);
        assert_eq!(live.lock().unwrap().len(), 1);
        assert_eq!(live.lock().unwrap()[&1].state, "paused");
        apply(&live, 1, row(1, "done", 10), true);
        assert!(live.lock().unwrap().is_empty());
    }

    #[test]
    fn the_list_reads_oldest_first_so_a_row_does_not_move_under_the_pointer() {
        let live = new_live_jobs();
        apply(&live, 2, row(2, "running", 200), false);
        apply(&live, 1, row(1, "running", 100), false);
        let map = live.lock().unwrap();
        let mut rows: Vec<JobRow> = map.values().cloned().collect();
        rows.sort_by(|a, b| a.started_at.cmp(&b.started_at).then(a.id.cmp(&b.id)));
        assert_eq!(
            rows.iter().map(|r| r.id.as_str()).collect::<Vec<_>>(),
            vec!["1", "2"]
        );
    }
}
